### PR TITLE
sql: implement restricted placement in zone configs

### PR DESCRIPTION
**Pull request by pawalt**
This PR adds support for the logic required to support restricted
placement i.e. removing non-voters from zone configs if restricted
placement is used.

Release note: None

### CODE REVIEW

File: pkg/sql/region_util.go [hunk 1]
```diff
@@ -375,11 +383,24 @@ func zoneConfigForMultiRegionTable(
 		// nothing here because `NewZoneConfig()` already marks those fields as
 		// 'inherited'.
 	case *descpb.TableDescriptor_LocalityConfig_RegionalByTable_:
-		// Use the same configuration as the database and return nil here.
+		restrictedPlacement := regionConfig.Placement() == descpb.DataPlacement_RESTRICTED
```
Comment (otan) [inline pkg/sql/region_util.go:386]: my comment from the earlier PR applies here; should we make the database descriptor have the restricted placement by default, and instead have global tables with DEFAULT placement strictness specify voters? this is prone to errors.

https://reviewable.io/reviews/cockroachdb/cockroach/67946#-MfGOVrlBMLqHltKbHRS

Comment (pawalt) [reply pkg/sql/region_util.go]: Agreed. Just pushed this up.

**Review by otan — COMMENTED** [review]
only nits from me, allowing arul to have final signoff!

> Reviewed 3 of 9 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68072#-:-MfeCr2b20-FAFvgiuv4:bxm44h4)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @otan, and @pawalt)

___
*[pkg/sql/region_util.go, line 200 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/68072#-MfeBotC7_9PlPMZe-tr:-MfeBotC7_9PlPMZe-ts:bpd1tq) ([raw file](https://github.com/cockroachdb/cockroach/blob/676d80ee90de44bb1f1ce35b27ecc5277079610c/pkg/sql/region_util.go#L200)):*
> ```Go
> 		{Constraints: []zonepb.Constraint{makeRequiredConstraintForRegion(partitionRegion)}},
> 	}
> ```

nit: revert
___
*[pkg/sql/region_util_test.go, line 659 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/68072#-MfeCa0YC_UBb2-Zouch:-MfeCa0YC_UBb2-Zouci:b-x8hdcv) ([raw file](https://github.com/cockroachdb/cockroach/blob/676d80ee90de44bb1f1ce35b27ecc5277079610c/pkg/sql/region_util_test.go#L659)):*
> ```Go
> 		},
> 		{
> 			desc: "4-region global table with restricted placement",
> ```

can you throw in a test for RBT and RBR as well?
___
*[pkg/sql/catalog/multiregion/region_config_test.go, line 48 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/68072#-MfeCgoTEIcqwCzaMI6v:-MfeCgoTEIcqwCzaMI6w:bohiz1n) ([raw file](https://github.com/cockroachdb/cockroach/blob/676d80ee90de44bb1f1ce35b27ecc5277079610c/pkg/sql/catalog/multiregion/region_config_test.go#L48)):*
> ```Go
> 		{
> 			err:          "expected > 0 number of regions in the region config",
> 			regionConfig: multiregion.MakeRegionConfig(descpb.RegionNames{}, "region_b", descpb.SurvivalGoal_REGION_FAILURE, validRegionEnumID, descpb.DataPlacement_DEFAULT),
> ```

nit: one arg per line like before (this makes the lines very long)

**Comment by otan** [comment]
thanks for applying that global table change!

**Review by ajstorm — COMMENTED** [review]
Lookin' good.  Would also be worth getting @aayushshah15 to have a peek, as he wrote the initial zone config stuff in 21.1.

> Reviewed 3 of 9 files at r1, 1 of 2 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68072#-:-Mfh7hD4CwpqiPY_XNCh:b-uy3x5u)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @otan, and @pawalt)

___
*[pkg/ccl/multiregionccl/multiregion.go, line 103 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/68072#-Mfh4WRODQyiqqKwevM8:-Mfh4WRODQyiqqKwevM9:b-ay7eyg) ([raw file](https://github.com/cockroachdb/cockroach/blob/676d80ee90de44bb1f1ce35b27ecc5277079610c/pkg/ccl/multiregionccl/multiregion.go#L103)):*
> ```Go
> 		// Using default placement for now until this gets stitched together with
> 		// the proper SQL statements.
> 		descpb.DataPlacement_DEFAULT,
> ```

Nit: I know in the initial proposal we called this DEFAULT, but I'm starting to wonder if we should give this a more descriptive name.  Maybe something like "EVERYWHERE" or "ALL_REGIONS" or something else?
___
*[pkg/sql/region_util.go, line 386 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/68072#-Mfh5Z_f75Z3TyTI6Wt7:-Mfh5Z_f75Z3TyTI6Wt8:bmq5wre) ([raw file](https://github.com/cockroachdb/cockroach/blob/676d80ee90de44bb1f1ce35b27ecc5277079610c/pkg/sql/region_util.go#L386)):*
> ```Go
> 
> 		// If we're in restricted placement, manually build the zone config so we
> 		// can get non-voters in all regions.
> ```

Nit: Can we beef up this comment?  I'm thinking something along the lines of:

```
// In cases where we're using RESTRICTED placement, we must build out the 
// voters/non-voters as well as the constraints fields.  This is because with RESTRICTED
// placement, the database-level zone configuration will be built to allow for regional
// tables to inherit the RESTRICTED placement policy.  For GLOBAL tables however, we
// need to place replicas in all regions to allow for fast reads in all regions.  
```

**Review by pawalt — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68072)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aayushshah15, @ajstorm, @arulajmani, and @otan)

___
*[pkg/ccl/multiregionccl/multiregion.go, line 103 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/68072#-Mfh4WRODQyiqqKwevM8:-MfhV9us8J8CRI6NQy3S:byb2dtj) ([raw file](https://github.com/cockroachdb/cockroach/blob/676d80ee90de44bb1f1ce35b27ecc5277079610c/pkg/ccl/multiregionccl/multiregion.go#L103)):*
<details><summary><i>Previously, ajstorm (Adam Storm) wrote…</i></summary><blockquote>

Nit: I know in the initial proposal we called this DEFAULT, but I'm starting to wonder if we should give this a more descriptive name.  Maybe something like "EVERYWHERE" or "ALL_REGIONS" or something else?
</blockquote></details>

Discussion in syntax PR
https://reviewable.io/reviews/cockroachdb/cockroach/68068
___
*[pkg/sql/region_util.go, line 386 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/68072#-Mfh5Z_f75Z3TyTI6Wt7:-Mfh_EFI-41V4wZEXFWG:b-942tt4) ([raw file](https://github.com/cockroachdb/cockroach/blob/676d80ee90de44bb1f1ce35b27ecc5277079610c/pkg/sql/region_util.go#L386)):*
<details><summary><i>Previously, ajstorm (Adam Storm) wrote…</i></summary><blockquote>

Nit: Can we beef up this comment?  I'm thinking something along the lines of:

```
// In cases where we're using RESTRICTED placement, we must build out the 
// voters/non-voters as well as the constraints fields.  This is because with RESTRICTED
// placement, the database-level zone configuration will be built to allow for regional
// tables to inherit the RESTRICTED placement policy.  For GLOBAL tables however, we
// need to place replicas in all regions to allow for fast reads in all regions.  
```
</blockquote></details>

Yup added a comment to that effect.
___
*[pkg/sql/region_util_test.go, line 659 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/68072#-MfeCa0YC_UBb2-Zouch:-MfhX7vd6PJgkYsr0AkB:b-1ysxcu) ([raw file](https://github.com/cockroachdb/cockroach/blob/676d80ee90de44bb1f1ce35b27ecc5277079610c/pkg/sql/region_util_test.go#L659)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

can you throw in a test for RBT and RBR as well?
</blockquote></details>

The return value for `zoneConfigForMultiRegionTable` doesn't change with PLACEMENT, so I don't think there's anything to test here. What we really would like to test is that the inheritance works properly for RBR and RBT, and I'll add logic tests for those in the PR that stitches the syntax and logic together.

**Review by arulajmani — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68072)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aayushshah15, @ajstorm, @otan, and @pawalt)

___
*[pkg/ccl/multiregionccl/multiregion.go, line 101 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/68072#-Mfj64KC887pAQfZgEfL:-Mfj64KC887pAQfZgEfM:b-lbabad) ([raw file](https://github.com/cockroachdb/cockroach/blob/db4ac4ca42fa6d2302133eddd9646fb1902eb540/pkg/ccl/multiregionccl/multiregion.go#L101)):*
> ```Go
> 		survivalGoal,
> 		regionEnumID,
> 		// Using default placement for now until this gets stitched together with
> ```

Add a TODO(peyton) before this? 
___
*[pkg/sql/catalog/descpb/structured.proto, line 1263 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/68072#-MfhkVxI6Kk5ETQcevXU:-MfhkVxI6Kk5ETQcevXV:bjsq7t0) ([raw file](https://github.com/cockroachdb/cockroach/blob/db4ac4ca42fa6d2302133eddd9646fb1902eb540/pkg/sql/catalog/descpb/structured.proto#L1263)):*
> ```protobuf
>     optional uint32 region_enum_id = 4 [(gogoproto.nullable) = false, (gogoproto.customname) = "RegionEnumID", (gogoproto.casttype) = "ID"];
> 
>     // Whether or not to use a restricted data placement policy.
> ```

nit: DataPlacement dictates ...
___
*[pkg/sql/catalog/multiregion/region_config.go, line 82 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/68072#-MfhmZQG9z-xU2MJAfSg:-MfhmZQG9z-xU2MJAfSh:b-7vkh1) ([raw file](https://github.com/cockroachdb/cockroach/blob/db4ac4ca42fa6d2302133eddd9646fb1902eb540/pkg/sql/catalog/multiregion/region_config.go#L82)):*
> ```Go
> 
> // Placement returns the data placement strategy.
> func (r *RegionConfig) Placement() descpb.DataPlacement {
> ```

Let's change this to `IsPlacementRestricted() bool` instead? I think it'll make all the places where you're comparing this to `descpb.RESTRICTED` clearer. 
___
*[pkg/sql/catalog/multiregion/region_config_test.go, line 41 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/68072#-MfhkrIb0bUFcjgvY9BY:-MfhkrIb0bUFcjgvY9BZ:bujb2z2) ([raw file](https://github.com/cockroachdb/cockroach/blob/db4ac4ca42fa6d2302133eddd9646fb1902eb540/pkg/sql/catalog/multiregion/region_config_test.go#L41)):*
> ```Go
> 				descpb.SurvivalGoal_ZONE_FAILURE,
> 				descpb.InvalidID,
> 				descpb.DataPlacement_DEFAULT),
> ```

nit: here, and below, could you put the `)` on a new line? 
___
*[pkg/sql/catalog/multiregion/region_config_test.go, line 63 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/68072#-Mfj4Sn2BpgpWaXu-aHP:-Mfj4Sn2BpgpWaXu-aHQ:b-8fa3g0) ([raw file](https://github.com/cockroachdb/cockroach/blob/db4ac4ca42fa6d2302133eddd9646fb1902eb540/pkg/sql/catalog/multiregion/region_config_test.go#L63)):*
> ```Go
> 				descpb.DataPlacement_DEFAULT),
> 		},
> 	}
> ```

Let's add a test case here for when you have both REGION SURVIVABILITY and Placement Restricted?

**Review by ajstorm — COMMENTED** [review]
Modulo getting Aayush's eyes on this, I'm at :lgtm: 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68072#-:-MfjxrMO9iyYKmCetUEN:b-op9fkh)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @aayushshah15, @ajstorm, @otan, and @pawalt)

**Review by pawalt — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68072)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @aayushshah15, @ajstorm, @arulajmani, and @otan)

___
*[pkg/ccl/multiregionccl/multiregion.go, line 101 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/68072#-Mfj64KC887pAQfZgEfL:-Mg6JGN17ydcMJBndvrZ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/db4ac4ca42fa6d2302133eddd9646fb1902eb540/pkg/ccl/multiregionccl/multiregion.go#L101)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Add a TODO(peyton) before this? 
</blockquote></details>

Done.
___
*[pkg/sql/catalog/multiregion/region_config.go, line 82 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/68072#-MfhmZQG9z-xU2MJAfSg:-Mg6KB2WDEu5pOsR5h9O:b858j7p) ([raw file](https://github.com/cockroachdb/cockroach/blob/db4ac4ca42fa6d2302133eddd9646fb1902eb540/pkg/sql/catalog/multiregion/region_config.go#L82)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Let's change this to `IsPlacementRestricted() bool` instead? I think it'll make all the places where you're comparing this to `descpb.RESTRICTED` clearer. 
</blockquote></details>

Sounds good. Also helps not leak out the `descpb` definitions into the code.
___
*[pkg/sql/catalog/multiregion/region_config_test.go, line 63 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/68072#-Mfj4Sn2BpgpWaXu-aHP:-Mg6Kr5H6cEq0RE0uE9a:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/db4ac4ca42fa6d2302133eddd9646fb1902eb540/pkg/sql/catalog/multiregion/region_config_test.go#L63)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Let's add a test case here for when you have both REGION SURVIVABILITY and Placement Restricted?
</blockquote></details>

Done.

**Review by arulajmani — APPROVED** [review]
:lgtm: 

> Reviewed 2 of 9 files at r1, 1 of 2 files at r2, 4 of 6 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68072#-:-MgBPl1R7P7dvvOj70fa:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @aayushshah15, @ajstorm, @arulajmani, @otan, and @pawalt)

___
*[pkg/sql/catalog/multiregion/region_config_test.go, line 87 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/68072#-MgBPXjZB1hg9lTubIWx:-MgBPXjZB1hg9lTubIWy:b-iyxepp) ([raw file](https://github.com/cockroachdb/cockroach/blob/163c88f728bb162600796c0503260d86001a60a1/pkg/sql/catalog/multiregion/region_config_test.go#L87)):*
> ```Go
> 			"expected err %v, got %v",
> 			tc.err,
> 			err)
> ```

nit: closing brace on the next line

**Review by aayushshah15 — APPROVED** [review]
:lgtm_strong: mod the nits about some of the comments. 

> Reviewed 3 of 9 files at r1, 1 of 7 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68072#-:-MgCJgTPET5RDdWOuhzd:bby7d0r)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 2 stale) (waiting on @aayushshah15, @ajstorm, @arulajmani, @otan, and @pawalt)

___
*[pkg/sql/region_util.go, line 145 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/68072#-MgC7pmIA2-7bJ1pR4YN:-MgC7pmIA2-7bJ1pR4YO:b-kd4px7) ([raw file](https://github.com/cockroachdb/cockroach/blob/e010e606b8ef376b76dd9bd8baf9ca65c6beadaa/pkg/sql/region_util.go#L145)):*
> ```Go
> 	numVoters, numReplicas := getNumVotersAndNumReplicas(regionConfig)
> 	var constraints []zonepb.ConstraintsConjunction
> 	if regionConfig.IsPlacementRestricted() {
> ```

Let's add a short comment here that explains why we have no constraints in the restricted case. Mention the fact that `voter_constraints` will be specified at the table/partition level, which will constrain all voters to the primary region.
___
*[pkg/sql/region_util.go, line 388 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/68072#-MgC8rkvE0sHJyou6--w:-MgC8rkvE0sHJyou6--x:b4ghi0p) ([raw file](https://github.com/cockroachdb/cockroach/blob/e010e606b8ef376b76dd9bd8baf9ca65c6beadaa/pkg/sql/region_util.go#L388)):*
> ```Go
> 		// placement. For GLOBAL tables, however, we want non-voters in all regions
> 		// for fast reads, so we have to manually build a zone config with the
> 		// nonvoters.
> ```

I think `zoneConfigForMultiRegionDatabase()` is a better place for the first part of this comment. 
___
*[pkg/sql/region_util.go, line 403 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/68072#-MgBrsKq0RT7EgCFuL0A:-MgBrsKq0RT7EgCFuL0B:btyx3a6) ([raw file](https://github.com/cockroachdb/cockroach/blob/e010e606b8ef376b76dd9bd8baf9ca65c6beadaa/pkg/sql/region_util.go#L403)):*
> ```Go
> 			// Placement only applies in zone survivability in which case voters are
> 			// only in the primary region. This means the total number of replicas is
> 			// 3 voting replicas + 1 for each non-primary region.
> ```

minor nit: let's use the predefined constant here, so: `s/3/numVotersForZoneSurvival`

**Review by pawalt — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68072)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 2 stale) (waiting on @aayushshah15, @ajstorm, @arulajmani, @otan, and @pawalt)

___
*[pkg/sql/region_util.go, line 145 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/68072#-MgC7pmIA2-7bJ1pR4YN:-MgGQXLB4LiY5N5Xo5-m:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e010e606b8ef376b76dd9bd8baf9ca65c6beadaa/pkg/sql/region_util.go#L145)):*
<details><summary><i>Previously, aayushshah15 (Aayush Shah) wrote…</i></summary><blockquote>

Let's add a short comment here that explains why we have no constraints in the restricted case. Mention the fact that `voter_constraints` will be specified at the table/partition level, which will constrain all voters to the primary region.
</blockquote></details>

Done.
___
*[pkg/sql/region_util.go, line 388 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/68072#-MgC8rkvE0sHJyou6--w:-MgGRCPgDASvsxNbOtK7:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e010e606b8ef376b76dd9bd8baf9ca65c6beadaa/pkg/sql/region_util.go#L388)):*
<details><summary><i>Previously, aayushshah15 (Aayush Shah) wrote…</i></summary><blockquote>

I think `zoneConfigForMultiRegionDatabase()` is a better place for the first part of this comment. 
</blockquote></details>

Done.

**Pull request merged**